storage/innobase/srv/srv0start.cc

```diff
@@ -509,13 +509,14 @@ static ulint srv_undo_tablespace_open(bool create, const char* name, ulint i)
   ulint n_retries = 5;
   os_offset_t size= os_file_get_size(fh);
   ut_a(size != os_offset_t(-1));
+  page_t *apage= nullptr;
+  const page_t *page= nullptr;
 
   if (!create)
   {
-    page_t *page= static_cast<byte*>(aligned_malloc(srv_page_size,
-                                                    srv_page_size));
+    apage= static_cast<byte*>(aligned_malloc(srv_page_size, srv_page_size));
 undo_retry:
-    if (os_file_read(IORequestRead, fh, page, 0, srv_page_size, nullptr) !=
+    if (os_file_read(IORequestRead, fh, apage, 0, srv_page_size, nullptr) !=
         DB_SUCCESS)
```

Comment:
Yes, we actually have to pass `page=apage` to `fil_node_t::read_page0()` in case there is no copy of the page in the doublewrite buffer. We do not know the final LSN at this point, so we may be using a "too new" copy of the page, either from this call or from the doublewrite buffer. We have no choice here; we must find all undo tablespaces and the system tablespace before we can start processing the redo log. This is because the redo log does not contain any `FILE_` records for these files.